File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
from collections import deque
import sys

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

n = int(sys.stdin.readline())
graph = []

for i in range(n):
    arr = list(map(int, sys.stdin.readline().rstrip()))
    graph.append(arr)

def bfs(graph, x, y):
    cnt = 0
    queue = deque([])
    queue.append([x, y])
    graph[x][y] = 0

    while queue:
        cnt += 1
        a, b = queue.popleft()

        for i in range(4):
            nx = dx[i] + a
            ny = dy[i] + b

            if nx < 0 or nx >= n or ny < 0 or ny >= n:
                continue
            if graph[nx][ny] == 1:
                queue.append([nx, ny])
                graph[nx][ny] = 0
    return cnt

ans = []
for i in range(n):
    for j in range(n):
        if (graph[i][j] == 1):
            ans.append(bfs(graph, i, j))

ans.sort()
print(len(ans))
for a in ans:
    print(a, sep='')

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];

rl.on("line", (input) => {
  lines.push(input.trim());
  if (lines.length === Number(lines[0]) + 1) {
    rl.close();
  }
});

rl.on("close", () => {
  const n = Number(lines[0]);
  const grid = [];
  for (let i = 1; i < n + 1; i++) {
    grid.push(lines[i].split("").map(Number));
  }

  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];

  function bfs(x, y) {
    const q = [];
    q.push([x, y]);
    grid[x][y] = 0;
    let cnt = 1;

    while (q.length) {
      const [x, y] = q.shift();
      for (let i = 0; i < 4; i++) {
        const nx = dx[i] + x,
          ny = dy[i] + y;
        if (0 <= nx && nx < n && 0 <= ny && ny < n) {
          if (grid[nx][ny] === 1) {
            grid[nx][ny] = 0;
            cnt++;
            q.push([nx, ny]);
          }
        }
      }
    }
    return cnt;
  }

  const answer = [];
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 1) {
        answer.push(bfs(i, j));
      }
    }
  }

  answer.sort((a, b) => a - b);
  console.log(answer.length);
  console.log(answer.join("\n"));
});
